cirq-core/cirq/linalg/operator_spaces.py

```diff
@@ -80,7 +80,7 @@ def matrix_from_basis_coefficients(
     some_element = next(iter(basis.values()))
     result = np.zeros_like(some_element, dtype=np.complex128)
     for name, coefficient in expansion.items():
-        result += coefficient * basis[name]
+        result += complex(coefficient) * basis[name]
```

Comment:
Relatedly I've been thinking that maybe parameter types should be split into two types, one being like a `SupportsParam / SupportsParamComplex` *protocol*, similar to e.g. `SupportsFloat`, and the other being our existing `ParamVal / ParamValComplex`, but where the union is one of exactly two options.

Then, gate constructors would accept a `Supports...` type for the parameter, but then convert it before setting the member variable. That allows more flexibility for instantiating gates, but more consistency when using them.

Reply:
Such an approach would also allow for an easy implementation of https://github.com/quantumlib/Cirq/issues/2936

Reply:
In this case the `result` array has been created with np.complex128 type whereas the `coefficient` from LinerDict value could be either a number or sympy expression causing an incompatible type on the RHS of `+=`.  Converting to complex makes the RHS type compatible.